lib/publish.js

```diff
@@ -1,167 +1,105 @@
-const libpub = require('libnpmpublish').publish
-const libunpub = require('libnpmpublish').unpublish
-const lifecycle = BB.promisify(require('./utils/lifecycle.js'))
+const util = require('util')
 const log = require('npmlog')
-const npa = require('npm-package-arg')
-const npmConfig = require('./config/figgy-config.js')
-const output = require('./utils/output.js')
-const otplease = require('./utils/otplease.js')
-const pack = require('./pack')
-const { tarball, extract } = require('pacote')
-const path = require('path')
-const readFileAsync = BB.promisify(require('graceful-fs').readFile)
-const readJson = BB.promisify(require('read-package-json'))
 const semver = require('semver')
-const statAsync = BB.promisify(require('graceful-fs').stat)
+const pack = require('libnpmpack')
+const libpub = require('libnpmpublish').publish
+const runScript = require('@npmcli/run-script')
+
+const npm = require('./npm')
+const output = require('./utils/output')
+const otplease = require('./utils/otplease')
+const { getContents, logTar } = require('./utils/tar')
 
-publish.usage = 'npm publish [<tarball>|<folder>] [--tag <tag>] [--access <public|restricted>] [--dry-run]' +
+const readJson = util.promisify(require('read-package-json'))
+
+publish.usage = 'npm publish [<folder>] [--tag <tag>] [--access <public|restricted>] [--dry-run]' +
                 "\n\nPublishes '.' if no argument supplied" +
                 '\n\nSets tag `latest` if no --tag specified'
 
-publish.completion = function (opts, cb) {
-  // publish can complete to a folder with a package.json
-  // or a tarball, or a tarball url.
-  // for now, not yet implemented.
-  return cb()
-}
-
-const PublishConfig = figgyPudding({
+const publishConfig = () => ({
   dryRun: 'dry-run',
-  'dry-run': { default: false },
-  force: { default: false },
-  json: { default: false },
-  Promise: { default: () => Promise },
-  tag: { default: 'latest' },
-  tmp: {}
+  tag: 'defaultTag',
+  'dry-run': false,
+  defaultTag: 'latest',
+  json: false,
+  tmp: {},
+  ...npm.flatOptions
 })
 
 module.exports = publish
-function publish (args, isRetry, cb) {
-  if (typeof cb !== 'function') {
-    cb = isRetry
-    isRetry = false
-  }
+async function publish (args, cb) {
   if (args.length === 0) args = ['.']
   if (args.length !== 1) return cb(publish.usage)
 
-  log.verbose('publish', args)
+  log.verbose('publish', args) 
 
-  const opts = PublishConfig(npmConfig())
-  const t = opts.tag.trim()
+  const opts = publishConfig()
+  const t = opts.defaultTag.trim()
   if (semver.validRange(t)) {
     return cb(new Error('Tag name must not be a valid SemVer range: ' + t))
   }
 
-  return publish_(args[0], opts)
-    .then((tarball) => {
-      const silent = log.level === 'silent'
-      if (!silent && opts.json) {
-        output(JSON.stringify(tarball, null, 2))
-      } else if (!silent) {
-        output(`+ ${tarball.id}`)
-      }
-    })
-    .nodeify(cb)
-}
+  const tarball = await publish_(args[0], opts)
```

Comment:
need a try/catch here to handle errors in the awaited promise.

Reply:
Disregard (and earlier comment to the same effect).  It'd be better to just use async and let it throw, and then have a top-level function that takes the callback and passes promise rejection to it.

```js
async function publish (args) { // no cb
  // do stuff
  // throw or return
  // await whatever
}

module.exports = (args, cb) => publish(args).then(() => cb()).catch(cb)
module.exports.usage = ...
module.exports.completion = ...
```